modules/tinymce/src/core/main/ts/api/EditorOptions.ts

```diff
@@ -0,0 +1,277 @@
+  const result = processor(value);
+  if (Type.isBoolean(result)) {
+    // Note: Need to cast here as if a boolean is returned then we're guaranteed to be returning the same value
+    return result ? { value: value as unknown as U, valid: true } : { valid: false, message: '' };
+  } else {
+    return result;
+  }
+};
+
+const processDefaultValue = <T, U>(name: string, defaultValue: T, processor: Processor<T, U>): U | undefined => {
+  if (!Type.isUndefined(defaultValue)) {
+    const result = processValue(defaultValue, processor);
+    if (isValidResult(result)) {
+      return result.value;
+    } else {
+      // eslint-disable-next-line no-console
+      console.error(getErrorMessage(`Invalid default value passed for the ${name} option`, result));
+    }
+  }
+
+  return undefined;
+};
+
+const create = (editor: Editor, initialOptions: Record<string, unknown>): Options => {
+  const registry: Record<string, OptionSpec<any, any>> = {};
+  const values: Record<string, any> = {};
+
+  editor.on('init', () => {
+    const unregisteredOptions = Arr.partition(Obj.keys(initialOptions), isRegistered).fail;
+    if (unregisteredOptions.length > 0) {
+      // eslint-disable-next-line no-console
+      console.warn('The following options were specified but have not been registered:\n - ' + unregisteredOptions.join('\n - '));
+    }
+  });
+
+  const setValue = <T, U>(name: string, value: T, processor: SimpleProcessor<T> | Processor<T, U>): boolean => {
+    const result = processValue(value, processor);
+    if (isValidResult(result)) {
+      values[name] = result.value;
+      // TODO: TINY-8236 (TINY-8234) Remove this later once all settings have been converted
+      editor.settings[name] = result.value;
+      return true;
+    } else {
+      // eslint-disable-next-line no-console
+      console.warn(getErrorMessage(`Invalid value passed for the ${name} option`, result));
+      return false;
+    }
+  };
+
+  const register = (name: string, spec: BuiltInOptionSpec<any> | OptionSpec<any, any>) => {
+    const processor = isBuiltInSpec(spec) ? getBuiltInProcessor(spec.processor) : spec.processor;
+
+    // Process and validate the default value
+    const defaultValue = processDefaultValue(name, spec.default, processor);
+
+    // Register the spec with the normalised validator
+    registry[name] = {
+      ...spec,
+      default: defaultValue,
+      processor
+    };
+
+    // Setup the initial values
+    if (!Obj.has(values, name)) {
+      if (!Type.isUndefined(defaultValue)) {
+        values[name] = defaultValue;
+      }
+      Obj.get(initialOptions, name).each((value) => setValue(name, value, processor));
+    }
+  };
+
+  const isRegistered = (name: string) =>
+    Obj.has(registry, name);
+
+  const get = (name: string) =>
+    Obj.get(values, name).getOrUndefined();
+
+  const set = <T>(name: string, value: T) => {
+    if (!isRegistered(name)) {
```

Comment:
I personally prefer non-negated if statements if possible, but not terribly fussed

Reply:
I did it this way because it's the way we conceptualise/walk through the process of what should happen.